samcli/lib/build/app_builder.py

```diff
@@ -428,28 +432,23 @@ def _build_layer(
 
             # By default prefer to build in-process for speed
             build_runtime = specified_workflow
-            build_method = self._build_function_in_process
+            options = ApplicationBuilder._get_build_options(layer_name, config.language, None)
             if self._container_manager:
-                build_method = self._build_function_on_container
                 if config.language == "provided":
                     LOG.warning(
                         "For container layer build, first compatible runtime is chosen as build target for container."
                     )
                     # Only set to this value if specified workflow is makefile
                     # which will result in config language as provided
                     build_runtime = compatible_runtimes[0]
-            options = ApplicationBuilder._get_build_options(layer_name, config.language, None)
+                self._build_function_on_container(
+                    config, code_dir, artifact_subdir, manifest_path, build_runtime, options, container_env_vars
+                )
+            else:
+                self._build_function_in_process(
+                    config, code_dir, artifact_subdir, scratch_dir, manifest_path, build_runtime, options
+                )
 
-            build_method(
```

Comment:
The refactoring makes sense here, since we are adding more and more container unique features, just wondering do we have make pr passing so far? I think this change might affect many test cases?

Reply:
Unit tests are passing. Integration tests have not been modified and not in the scope of this PR.

Reply:
> Integration tests have not been modified and not in the scope of this PR.

I am concerned that we are splitting up prs on testing. This will make it much harder to understand changes. I know this PR is going into a feature branch but we lose sight of what is actually changing and what needs to be added later. I would really discourage this behavior and push us to have the context within the same pr. There can be times for this, but it should not be the typical process we follow. If this was a contribution from a community member, we would be asking them to add the integ tests into this pr. Why are we treating ourselves in a different manner?

Reply:
> > Integration tests have not been modified and not in the scope of this PR.
> 
> I am concerned that we are splitting up prs on testing. This will make it much harder to understand changes. I know this PR is going into a feature branch but we lose sight of what is actually changing and what needs to be added later. I would really discourage this behavior and push us to have the context within the same pr. There can be times for this, but it should not be the typical process we follow. If this was a contribution from a community member, we would be asking them to add the integ tests into this pr. Why are we treating ourselves in a different manner?

This was done for this specific PR because of capacity reasons and trying to make implementation and tests parallel between Qingchuan and me.